### PR TITLE
Add `File.readLine` for buffered IO

### PR DESCRIPTION
**Pull request by lukewilliamboswell**

### CODE REVIEW

**Comment by Anton-4** [comment]
One thing I'd like to improve is that `readLine` and `fileReadLine` return an `Err` on EOF. But EOF is expected in a normal flow, not truly an error. I'd like to return a tag union on `Ok` but I could not find any examples where we return Roc tag union in Rust code, do you know if this is possible @lukewilliamboswell?

**Comment by lukewilliamboswell** [comment]
> One thing I'd like to improve is that `readLine` and `fileReadLine` return an `Err` on EOF. But EOF is expected in a normal flow, not truly an error. I'd like to return a tag union on `Ok` but I could not find any examples where we return Roc tag union in Rust code, do you know if this is possible @lukewilliamboswell?

That will be easy to add. Just want to clarify a couple of things @Anton-4 ;
- Can we just return an empty list for EOF? then the API stays `readLine : FileReader -> Task (List U8) [FileReadErr Path Str]`. I suspect this will be the easiest to work with in roc.
- If we want to change it, what tag union would you like? how about `Task [BytesRead (List U8), EOF] [FileReadErr Path Str]`?

**Comment by Anton-4** [comment]
> Can we just return an empty list for EOF?

Interesting idea, let's try it out :)

File: platform/File.roc [hunk 1]
```diff
@@ -181,3 +184,45 @@ isSymLink = \path ->
...
+FileReader := {readerID: U64, path: Path}
+
+## Try to create a `FileReader` for buffered (= part by part) reading given a path string.
+## See [examples/file-read-buffered.roc](https://github.com/roc-lang/basic-cli/blob/main/examples/file-read-buffered.roc) for example usage.
+##
+## This uses [rust's std::io::BufReader](https://doc.rust-lang.org/std/io/struct.BufReader.html).
+##
+## Use [readUtf8] if you want to get the entire file contents at once.
+getFileReader : Str -> Task FileReader [FileReadErr Path ReadErr]
+getFileReader = \pathStr ->
```
Comment (smores56) [inline platform/File.roc:198]: There doesn't seem to be a way to close a file, is that a goal for this PR?

Comment (lukewilliamboswell) [reply platform/File.roc:198]: That's a really good point! 😓

Comment (lukewilliamboswell) [reply platform/File.roc:199]: Done, added a `File.closeFileReader` effect. It closes the file, and then any future "reads" will simply return NIL bytes.

File: platform/src/lib.rs [hunk 2]
```diff
@@ -545,6 +590,66 @@ pub extern "C" fn roc_fx_fileReadBytes(
...
+
+        let dummy_reader = BufReader::new(std::io::Cursor::new(Vec::new()));
+
+        // we drop the BufReader<File> to manually close the file
+        // by replacing it with a dummy reader that will return EOF
+        // for any future read attempts.
+        *reader_ref_cell.borrow_mut() = Box::new(dummy_reader);
+
+    })
+}
```
Comment (lukewilliamboswell) [inline platform/src/lib.rs:651]: @smores56 what do you think of this? 

I'm a rust noob, but this seems to be working ok. (segfault on linux aside -- which seems unrelated)

File: platform/src/lib.rs [hunk 3]
```diff
@@ -545,6 +559,66 @@ pub extern "C" fn roc_fx_fileReadBytes(
...
+
+        let reader_ref_cell = readers.get(readerIndex as usize).expect("readerIndex was larger than number of readers in Vec. We expect this can only happen due to an implementation error in basic-cli.");
+
+        let mut string_buffer = String::new();
+
+        let mut file_buf_reader = reader_ref_cell.borrow_mut();
+
+        match file_buf_reader.read_line(&mut string_buffer) {
+            Ok(bytes_read) => {
+                if bytes_read == 0 {
```
Comment (smores56) [inline platform/src/lib.rs:593]: It seems like even without this clause the `string_buffer` would be empty, so it's probably unnecessary. I could understand keeping it for readability reasons.

File: platform/src/lib.rs (hunk 2 above)
Comment (smores56) [reply platform/src/lib.rs:620]: My instinct is to say we should just remove it from the READERS list, but that would break the indices of the other readers. So unless we change this to a `HashMap`, or a ring-buffer of some sort, the current way this works is fine. It will just grow over time, which is sub-optimal.

File: platform/src/lib.rs (hunk 3 above)
Comment (lukewilliamboswell) [reply platform/src/lib.rs]: Ah, thank you.

File: platform/src/lib.rs (hunk 2 above)
Comment (lukewilliamboswell) [reply platform/src/lib.rs]: Done, switch to a `HashMap` which is much simpler.

**Review by smores56 — APPROVED** [review]

File: examples/file-read-buffered.roc [hunk 4]
```diff
@@ -23,20 +23,13 @@ main =
...
-    Stdout.line! "Done reading, got $(Inspect.toStr readSummary)"
-
-    # If we close a reader, we will get EOF for any future reads.
-    closedReader = File.getFileReader! "README.md"
-
-    File.closeFileReader! closedReader
-
-    when File.readLine closedReader |> Task.result! is
-        Ok bytes if List.len bytes == 0 -> Stdout.line "Got EOF reading from closed file."
-        _ -> crash "unexpected; we should get an EOF trying to read a file after closing"
```
Comment (Anton-4) [inline examples/file-read-buffered.roc]: What do you think about this change @lukewilliamboswell?
I feel like the examples should briefly demonstrate best practices, explaining this edge case does not seem necessary.

**Review by Anton-4 — APPROVED** [review]

**Comment by Anton-4** [comment]
Thanks to everyone involved :)

This is good to merge for me when Luke is [ok with this](https://github.com/roc-lang/basic-cli/pull/206#discussion_r1654934126).

**Comment by lukewilliamboswell** [comment]
Thank you @Anton-4, looks good. I guess I added that extra bit to help me validate it was actually closing the file. I was suspect of that just `Drop`ing the handle would close it and wanted to QA that manually. But it's no longer required for an example.
